### ISSUE TITLE
The meaning of `isTrainable` property of a layer

### ISSUE THREAD

**Issue by mkaze** [JetBrains/KotlinDL#152]
I am a bit confused about the meaning of `isTrainable` property of layers, so I decided to create this issue to ask for clarification.

In Keras, there is a boolean `trainable` attribute for each layer which indicates whether the weights of a layer (if it has any) should be updated during training or not (i.e. whether the layer is freezed or not during training). This is always set to `True` upon creation of the layer (even if the layer has no weights), unless the user explicitly set it to `False`. Now, if I am correct, its equivalent in KotlinDL is `isTrainable` property. From its docs (although, there is a typo: "... and could **not** be changed ..."):

> True, if layer's weights could be changed during training. If false, layer's weights are frozen and could be changed during the training.

But for the following layers, this property has been set to `false`:
- `AbstractActivationLayer` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/activation/AbstractActivationLayer.kt#L29
- `BatchNorm` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/normalization/BatchNorm.kt#L63 https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/normalization/BatchNorm.kt#L77
- `Concatenate` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/merge/Concatenate.kt#L26
- `Conv3D` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/Conv3D.kt#L96
- `Dropout` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/regularization/Dropout.kt#L36
- `DepthwiseConv2D` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/DepthwiseConv2D.kt#L92
- `SeparableConv2D` https://github.com/JetBrains/KotlinDL/blob/03c18901af9dc86edbf0e0d1d199e787bdc2a9dd/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/SeparableConv2D.kt#L102

This is a bit confusing to me, especially for layers with trainable weights (e.g. `Conv3D`). Am I missing something, or is this a bug? @zaleslaw

**Comment by zaleslaw** [comment]
Hi, this is a good question:

1. You found a typo; it should be "and could not be changed."
2. The problem is the following: only part of TF ops have gradients (implemented [here](https://github.com/tensorflow/tensorflow/tree/r1.15/tensorflow/cc/gradients)) and could be used during training, a few layers like DepthwiseConv2D, Concatenate, SeparableConv2D, and so on. 
3. Some layers like BatchNorm (or Dropout) have no implementation for "training"; for inference only, weights could be loaded but could not be updated.
4. It could be the same for many new layers added at this moment by you and other community members; we will explore it during reading the sources or getting an error from low-level C API about nonexisting gradients if we try to update the weights in some architectures.
5. Currently TF community is working on new gradients, and probably we obtain them in the next 1-2 years.

**Comment by mkaze** [comment]
Interesting, weird and confusing at the same time! All these years I thought the Python API of TF is just a high-level wrapper around its C++ core implementation, or at least most of its functionalities are delegated to its C++ core implementation. Now, it's a surprise for me to see that it's not actually the case and something like gradients are not fully supported in C++ implementation and are being ported from Python!

**Comment by dosier** [comment]
What about the merge, activation and reshaping layers? I take all of those should also have `isTrainable == false`?

**Comment by mkaze** [comment]
@zaleslaw 

> 5. Currently TF community is working on new gradients, and probably we obtain them in the next 1-2 years.

I just read [this RFC](https://github.com/tensorflow/community/blob/master/rfcs/20201201-cpp-gradients.md), but I am a bit confused. Just to get a better understanding of the situation for myself, as far as I know:
 - all of the TF APIs in languages other than C/C++ (e.g. Python, Java, etc.), at the end are basically calling the C/C++ code under the hood,
 - and all the **basic ops** in Python or Java API of TF are already implemented and supported (which again delegate to C/C++ code),

then is it correct to say that for a more complex op which is currently missing we should be able to directly use its corresponding implementation in C/C++ API or, if there is no such implementation, implement that op using the existing basic ops? 

As an example, suppose we want to implement Softmax Gradient op in Java API (let's assume it's missing). Then if there is such an op in C/C++ API we can directly use that, or it's not in the C++ API, we can implement it in Java using the existing basic ops/functionality in TF Java API. Is that a correct description of the situation? 

Or to put it differently: if some op or functionality, like gradient tape, exists in TF Python API, then what's stopping us to implement it in TF Java API the same way it has been implemented in Python? Wouldn't they both use the C/C++ API directly or indirectly, after all?

**Comment by zaleslaw** [comment]
Hi, @mkaze great to see you again
I'll try to answer all of your questions and leave a few comments from my side:

1. You mentioned the [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20201201-cpp-gradients.md). yes, it exists, but not finished and merged into the codebase

2. This summer, a few new gradients were added to the [master branch](https://github.com/tensorflow/tensorflow/commit/bf3d89b1e8bc5e00271c2fbd57a8b3ff55d48462#diff-cc5eff4cd37583c8fcba5ec6a9c843d2acf4e31c8f71c20aacb6c3fe57e601bb) for ConcatV2 and FusedBatchNorm ops, for example. I hope that these gradients will be available for Java development via [Java API for TF](https://github.com/tensorflow/java) and later will be available for all Java API for TF users like KotlinDL

3. Java (and Kotlin as a result) has a way to call C style methods via JNI. I don't know how it works for Python, but C API for TF, in reality, does not cover all the possibilities of С++ API. It makes me cry, but on the other hand, the C API is supposed to serve officially as the only "stable" interface to the TensorFlow runtime, not C++, and all Java APIs (1.15 used in KotlinDL and 2.x) rely on it.

4. To add your own gradients in a client code like in [Python ](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/ops/nn_grad.py) via ```@tf.RegisterGradient```, you need access to the special place in C API exposed to the Java API to register the gradients, but this is missed or doesn't work

5. [This](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/framework/ops.py#L2675) is how it's implemented in Python, and the Python contains the whole framework for handling ConcreteFunctions, GradientTape, Gradients after registration, and probably it should be a port to the Java/Kotlin to register and maintain custom gradients.

5. Read this [discussion ](https://github.com/tensorflow/community/pull/335)( I tried to find the same way a year ago)

I spent a lot of time figuring out how to solve this problem but not finding a good solution that could solve the problem in a reasonable amount of time, so I decided to move forward, focused on inference and transfer-learning for TensorFlow.

If you are really interested in this topic, we could arrange a call to discuss it.

**Comment by mkaze** [comment]
@zaleslaw Thank you for your reply and explanations. I appreciate it.

But still I don't understand why the coverage of TF Java API is limited compared to TF Python API? I mean is there any specific obstacle/limitation related to Java or the underlying C API of TF, or is it only because no one has invested enough time and effort to extend the TF Java API?

**Comment by zaleslaw** [comment]
Java API for TF used auto-generation to generate Java classes for exposed C API, and it regenerates from time to time (in the case of [Java API for TF 2.x](https://github.com/tensorflow/java))

This corresponds to the ops generated in python, [gen_nn_ops ](https://github.com/tensorflow/tensorflow/blob/cf19edcdc547dc71f3a22c1f87247e8881adacd9/tensorflow/python/ops/nn_ops.py#L506) for example.

The next task - to implement the low-level logic written in python just in this [package](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/ops), and on this level a lot of missed things.

Gen_nn_ops, for example, are generated in the same way in [Java ](https://github.com/tensorflow/java/tree/master/tensorflow-core/tensorflow-core-api/src/bazel/api_def) and in [Python](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/api_def/python_api), via api-defs, in both cases you get access to the underlying [C++ code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/nn_ops.cc), I hope.

If we will talk about probable obstacles/limitations, the honest answer is "I don't know". it looks like there are no obstacles, but I could not guarantee that.
